pkg/api/v1/ports_update.go

```diff
@@ -53,20 +54,167 @@ func (r *Router) portUpdate(c echo.Context) error {
 
-	if err := validatePort(port); err != nil {
+	portID, err := r.updatePort(c, port, origPools, payload.Pools)
+	if err != nil {
+		return err
+	}
+
+	return v1UpdatePortResponse(c, portID)
+}
+
+// portPatch patches a port
+func (r *Router) portPatch(c echo.Context) error {
+	ctx := c.Request().Context()
+
+	mods, err := r.portParamsBinding(c)
+	if err != nil {
+		r.logger.Error("failed to bind port params", zap.Error(err))
 		return v1BadRequestResponse(c, err)
 	}
 
-	if _, err := port.Update(ctx, r.db, boil.Infer()); err != nil {
-		r.logger.Error("failed to update port", zap.Error(err))
+	ports, err := models.Ports(mods...).All(ctx, r.db)
+	if err != nil {
+		r.logger.Error("failed to get port", zap.Error(err))
 		return v1InternalServerErrorResponse(c, err)
 	}
 
+	if len(ports) == 0 {
+		return v1NotFoundResponse(c)
+	} else if len(ports) != 1 {
+		return v1BadRequestResponse(c, ErrAmbiguous)
+	}
+
+	port := ports[0]
+
+	// collect original pool ids for the port
+	origPools := make([]string, len(port.R.Assignments))
+	for i, p := range port.R.Assignments {
+		origPools[i] = p.PoolID
+	}
+
+	payload := struct {
+		Name  *string  `json:"name"`
+		Port  *int64   `json:"port"`
+		Pools []string `json:"pools"`
+	}{}
+
+	if err := c.Bind(&payload); err != nil {
+		r.logger.Error("failed to bind port update input", zap.Error(err))
+		return v1BadRequestResponse(c, err)
+	}
+
+	if payload.Name != nil {
+		port.Name = *payload.Name
+	}
+
+	if payload.Port != nil {
+		port.Port = *payload.Port
+	}
+
+	if payload.Pools == nil {
+		payload.Pools = origPools
+	}
+
+	// TODO do we need to update a CurrentState here?
+
+	portID, err := r.updatePort(c, port, origPools, payload.Pools)
+	if err != nil {
+		return err
+	}
+
+	return v1UpdatePortResponse(c, portID)
+}
+
+func (r *Router) updatePort(c echo.Context, port *models.Port, origPools, newPools []string) (string, error) {
+	ctx := c.Request().Context()
+
+	if err := validatePort(port); err != nil {
+		return "", v1BadRequestResponse(c, err)
+	}
+
+	// validate load balancer
+	lb, err := models.LoadBalancers(
+		models.LoadBalancerWhere.LoadBalancerID.EQ(port.LoadBalancerID),
+	).One(ctx, r.db)
+	if err != nil {
+		r.logger.Error("error looking up load balancer for port", zap.Error(err))
+		return "", v1BadRequestResponse(c, err)
+	}
+
+	tx, err := r.db.BeginTx(ctx, nil)
+	if err != nil {
+		r.logger.Error("failed to begin transaction", zap.Error(err))
+		return "", v1InternalServerErrorResponse(c, err)
+	}
+
+	if _, err := port.Update(ctx, tx, boil.Infer()); err != nil {
+		r.logger.Error("failed to update port", zap.Error(err))
+
+		if err := tx.Rollback(); err != nil {
+			r.logger.Error("error rolling back transaction", zap.Error(err))
+			return "", v1InternalServerErrorResponse(c, err)
+		}
+
+		return "", v1InternalServerErrorResponse(c, err)
+	}
+
+	additionalURNs := []string{
+		pubsub.NewLoadBalancerURN(lb.LoadBalancerID),
+	}
+
+	// add new assignments
```

Comment:
you could make finding the diff b/w the updates more performant by using a map
https://play.golang.com/p/j4tBOadvo6x

Reply:
good suggestion!